lib/pages/chat_search/chat_search.dart

```diff
@@ -41,38 +41,60 @@ class ChatSearchController extends State<ChatSearch> {
 
   final inputFocus = FocusNode();
 
-  final debouncer = Debouncer(
-    _debouncerDuration,
-    initialValue: '',
-    checkEquality: false,
-  );
+  SameTypeEventsBuilderController? sameTypeEventsBuilderController;
 
-  SameTypeEventsBuilderController? eventsController;
+  ServerSearchController? serverSearchController;
 
   final scrollController = ScrollController();
 
   @override
   void initState() {
-    eventsController = SameTypeEventsBuilderController(
-      getTimeline: getTimeline,
-      searchFunc: (event) =>
-          event.isContains(debouncer.value) && event.isSearchable,
-      limit: _pageLimit,
-    );
-    scrollController.addLoadMoreListener(eventsController!.loadMore);
+    if (room?.encrypted == true) {
+      sameTypeEventsBuilderController = SameTypeEventsBuilderController(
+        getTimeline: getTimeline,
+        searchFunc: (event) =>
+            event.isContains(debouncerValue) && event.isSearchable,
+        limit: _pageLimit,
+      );
+    } else {
+      serverSearchController = ServerSearchController();
+    }
+    scrollController.addLoadMoreListener(loadMore);
     textEditingController.addListener(() {
-      debouncer.value = textEditingController.text;
+      setDebouncerValue(textEditingController.text);
     });
-    debouncer.values.listen((text) {
-      if (text.length >= AppConfig.chatRoomSearchKeywordMin) {
-        eventsController?.refresh(force: true);
+    initializeDebouncer((text) {
+      if (sameTypeEventsBuilderController != null) {
+        if (text.length >= AppConfig.chatRoomSearchKeywordMin) {
+          sameTypeEventsBuilderController?.refresh(force: true);
+        } else {
+          sameTypeEventsBuilderController?.clear();
+        }
       } else {
-        eventsController?.clear();
+        if (text.isNotEmpty) {
+          serverSearchController?.search(
+            text,
+            roomId: widget.roomId,
+            onDone: () {
+              scrollController.tryLoadMoreIfNeeded(loadMore);
+            },
+          );
+        } else {
+          serverSearchController?.clear();
+        }
```

Comment:
not really understand, can you find other approach to do? method? abstraction? manager? ....